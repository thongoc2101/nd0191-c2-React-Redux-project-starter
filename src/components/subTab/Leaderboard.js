import { useSelector } from "react-redux";
import React from "react";

const Leaderboard = () => {
    const { users } = useSelector((state) => state.auth);

    return (
        <div className="p-6">
            <table className="w-full text-sm">
                <thead className="table-auto">
                    <tr className="table-row">
                        <th className="border border-solid bg-gray-200 text-left pl-4">Users</th>
                        <th className="border border-solid bg-gray-200">Answered</th>
                        <th className="border border-solid bg-gray-200">Created</th>
                    </tr>
                </thead>
                <tbody>
                    {/* Map over the user info and generate table rows */}
                    {Object.values(users)
                        ?.sort((a, b) => (
                            (Object.values(b.answers).length + b.questions.length) - (Object.values(a.answers).length + a.questions.length)))
                        .map((user) => (
                            <tr key={user.id}>
                                <td className="border border-solid text-left pl-4">
                                    <div className="text-sm font-bold">{user.name}</div>
                                    <div className="text-xs">{user.id}</div>
                                </td>
                                <td className="border border-solid">{Object.values(user.answers).length}</td>
                                <td className="border border-solid">{user.questions.length}</td>
                            </tr>
                        ))}
                </tbody>
            </table>
        </div>
    );
}

export default Leaderboard;