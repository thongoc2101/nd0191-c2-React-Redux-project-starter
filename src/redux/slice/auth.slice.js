import { createSlice } from "@reduxjs/toolkit";
import { AuthActions } from "../actions/auth.action";

const initialState = {
    userInfo: null,
    users: null,
    isLogin: false
};

export const AuthSlice = createSlice({
    name: 'auth',
    initialState: initialState,
    reducers: {
        login: (state, action) => {
            const user = Object.values(state.users).find(user => user.id === action.payload.username && user.password === action.payload.password);
            if (user) {
                state.isLogin = true;
                state.userInfo = user;
            } else {
                state.isLogin = false;
                state.userInfo = null;
            }
        },
        logout: (state) => {
            state.isLogin = false;
            state.userInfo = null;
        },

        setQuestionUser: (state, action) => {
            const userItem = Object.values(state.users).find(user => user.id === action.payload.author);
            userItem.questions.push(action.payload.id);

            state.userInfo = userItem;
            state.users = {
                ...state.users,
                [userItem.id]: userItem
            }
        },

        setUserAnswer: (state, action) => {
            const authedUser = action.payload.authedUser;
            const answer = action.payload.answer;
            const qid = action.payload.id;

            state.users = {
                ...state.users,
                [authedUser]: {
                  ...state.users[authedUser],
                  answers: {
                    ...state.users[authedUser].answers,
                    [qid]: answer
                  }
                }
              }
        }
    },
    extraReducers: (builder) => {
        // Add extra reducers here
        builder.addCase(AuthActions.getUsers.fulfilled, (state, action) => {
            state.users = action.payload;
        });
    }
});

export const { login, logout, setQuestionUser, setUserAnswer } = AuthSlice.actions;