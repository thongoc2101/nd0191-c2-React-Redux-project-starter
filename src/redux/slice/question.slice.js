import { createSlice } from "@reduxjs/toolkit";
import { QuestionActions } from "../actions/question.action";

const initialState = {
    questions: null,
};

export const QuestionSlice = createSlice({
    name: 'questions',
    initialState: initialState,
    reducers: {
        setAnsweredUser: (state, action) => {
            const qid = action.payload.id;
            const answer = action.payload.answer;
            const authedUser = action.payload.authedUser;

            state.questions = {
                ...state.questions,
                [qid]: {
                  ...state.questions[qid],
                  [answer]: {
                    ...state.questions[qid][answer],
                    votes: state.questions[qid][answer].votes.push(authedUser)
                  }
                }
              }
        }
    },
    extraReducers: (builder) => {
        // Add extra reducers here
        builder.addCase(QuestionActions.getQuestions.fulfilled, (state, action) => {
            state.questions = action.payload;
        });
    }
});

export const { setAnsweredUser } = QuestionSlice.actions;