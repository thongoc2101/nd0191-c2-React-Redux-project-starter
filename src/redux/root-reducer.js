import { combineReducers } from "@reduxjs/toolkit";
import { AuthSlice } from "./slice/auth.slice";
import { QuestionSlice } from "./slice/question.slice";

export const rootReducer = combineReducers({
    auth: AuthSlice.reducer,
    questions: QuestionSlice.reducer,
});