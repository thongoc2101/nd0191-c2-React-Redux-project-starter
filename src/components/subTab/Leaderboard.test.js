import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import {store} from "../../redux/store";
import Leaderboard from './Leaderboard';
import { AuthActions } from '../../redux/actions/auth.action';
import { login } from '../../redux/slice/auth.slice';

describe("Leaderboard", () => {
    it("should render the component when logged in", async () => {
        await store.dispatch(AuthActions.getUsers());
        await store.dispatch(login({ username: 'tylermcginnis', password: 'abc321'}))

        const component = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Leaderboard />
                </BrowserRouter>
            </Provider>
        );

        expect(component).toBeDefined();
        expect(component).toMatchSnapshot();
    });
});
