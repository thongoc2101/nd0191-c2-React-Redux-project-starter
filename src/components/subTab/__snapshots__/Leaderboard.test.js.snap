// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Leaderboard should render the component when logged in 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="p-6"
      >
        <table
          class="w-full text-sm"
        >
          <thead
            class="table-auto"
          >
            <tr
              class="table-row"
            >
              <th
                class="border border-solid bg-gray-200 text-left pl-4"
              >
                Users
              </th>
              <th
                class="border border-solid bg-gray-200"
              >
                Answered
              </th>
              <th
                class="border border-solid bg-gray-200"
              >
                Created
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td
                class="border border-solid text-left pl-4"
              >
                <div
                  class="text-sm font-bold"
                >
                  Sarah Edo
                </div>
                <div
                  class="text-xs"
                >
                  sarahedo
                </div>
              </td>
              <td
                class="border border-solid"
              >
                4
              </td>
              <td
                class="border border-solid"
              >
                2
              </td>
            </tr>
            <tr>
              <td
                class="border border-solid text-left pl-4"
              >
                <div
                  class="text-sm font-bold"
                >
                  Mike Tsamis
                </div>
                <div
                  class="text-xs"
                >
                  mtsamis
                </div>
              </td>
              <td
                class="border border-solid"
              >
                3
              </td>
              <td
                class="border border-solid"
              >
                2
              </td>
            </tr>
            <tr>
              <td
                class="border border-solid text-left pl-4"
              >
                <div
                  class="text-sm font-bold"
                >
                  Tyler McGinnis
                </div>
                <div
                  class="text-xs"
                >
                  tylermcginnis
                </div>
              </td>
              <td
                class="border border-solid"
              >
                2
              </td>
              <td
                class="border border-solid"
              >
                2
              </td>
            </tr>
            <tr>
              <td
                class="border border-solid text-left pl-4"
              >
                <div
                  class="text-sm font-bold"
                >
                  Zenobia Oshikanlu
                </div>
                <div
                  class="text-xs"
                >
                  zoshikanlu
                </div>
              </td>
              <td
                class="border border-solid"
              >
                1
              </td>
              <td
                class="border border-solid"
              >
                0
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="p-6"
    >
      <table
        class="w-full text-sm"
      >
        <thead
          class="table-auto"
        >
          <tr
            class="table-row"
          >
            <th
              class="border border-solid bg-gray-200 text-left pl-4"
            >
              Users
            </th>
            <th
              class="border border-solid bg-gray-200"
            >
              Answered
            </th>
            <th
              class="border border-solid bg-gray-200"
            >
              Created
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td
              class="border border-solid text-left pl-4"
            >
              <div
                class="text-sm font-bold"
              >
                Sarah Edo
              </div>
              <div
                class="text-xs"
              >
                sarahedo
              </div>
            </td>
            <td
              class="border border-solid"
            >
              4
            </td>
            <td
              class="border border-solid"
            >
              2
            </td>
          </tr>
          <tr>
            <td
              class="border border-solid text-left pl-4"
            >
              <div
                class="text-sm font-bold"
              >
                Mike Tsamis
              </div>
              <div
                class="text-xs"
              >
                mtsamis
              </div>
            </td>
            <td
              class="border border-solid"
            >
              3
            </td>
            <td
              class="border border-solid"
            >
              2
            </td>
          </tr>
          <tr>
            <td
              class="border border-solid text-left pl-4"
            >
              <div
                class="text-sm font-bold"
              >
                Tyler McGinnis
              </div>
              <div
                class="text-xs"
              >
                tylermcginnis
              </div>
            </td>
            <td
              class="border border-solid"
            >
              2
            </td>
            <td
              class="border border-solid"
            >
              2
            </td>
          </tr>
          <tr>
            <td
              class="border border-solid text-left pl-4"
            >
              <div
                class="text-sm font-bold"
              >
                Zenobia Oshikanlu
              </div>
              <div
                class="text-xs"
              >
                zoshikanlu
              </div>
            </td>
            <td
              class="border border-solid"
            >
              1
            </td>
            <td
              class="border border-solid"
            >
              0
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
